### PR TITLE
[SPARK-41358][SQL] Refactor `ColumnVectorUtils#populate` method to use `PhysicalDataType` instead of `DataType` 

### PR DESCRIPTION
**Pull request by LuciferYang**
### What changes were proposed in this pull request?
The main change of this pr is refactor `ColumnVectorUtils#populate` method to use `PhysicalDataType` instead of `DataType`.

### Why are the changes needed?
Simplify type match.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing ColumnVectorUtilsSuite

### CODE REVIEW

**Comment by LuciferYang** [comment]
@cloud-fan @desmondcheongzx I create an Umbrella SPARK-41356 and add 2 possible initial tasks and will add subTasks gradually in the future, If you have time, please help to review the first one. Thanks ~

File: sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+package org.apache.spark.sql.types;
+
+import org.apache.spark.sql.catalyst.types.*;
+
+/**
+ * To get specific physical data type.
+ *
+ * @since 3.4.0
+ */
+public class PhysicalDataTypes {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java:27]: is this really needed? `PhysicalDataType` is an internal API, and we don't need to make it java friendly. We can write `df isinstanceof PhysicalStringType`.

Comment (LuciferYang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java:27]: For example, in the following scenario:

https://github.com/apache/spark/blob/3fc8a90267312fdfc42b7e6a16ee69f2507ef56b/sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/UnsafeRow.java#L79-L95

if we can replace `PhysicalDataType` with `DataType`:

```java
 mutableFieldTypes = Collections.unmodifiableSet(
      new HashSet<>(
        Arrays.asList(
          PhysicalNullType$.MODULE$,
          PhysicalBooleanType$.MODULE$,
          PhysicalByteType$.MODULE$,
          PhysicalShortType$.MODULE$,
          PhysicalIntegerType$.MODULE$,
          PhysicalLongType$.MODULE$,
          PhysicalFloatType$.MODULE$,
          PhysicalDoubleType$.MODULE$
        )));
```

Does it look ugly

Comment (LuciferYang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java:27]: If it's not ugly, I think `PhysicalDataTypes` can be deleted

Comment (cloud-fan) [reply sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java:27]: It's OK. This is internal code anyway.

Comment (LuciferYang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java:27]: OK, let me remove PhysicalDataTypes and change to use `isinstanceof `

Comment (LuciferYang) [reply sql/catalyst/src/main/java/org/apache/spark/sql/types/PhysicalDataTypes.java]: done

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java [hunk 2]
```diff
@@ -125,32 +122,37 @@ public static Map<Integer, Integer> toJavaIntMap(ColumnarMap map) {
...
         dst.appendByte((Byte) o);
-      } else if (t == DataTypes.ShortType) {
+      } else if (pdt instanceof PhysicalShortType) {
         dst.appendShort((Short) o);
-      } else if (t == DataTypes.IntegerType) {
-        dst.appendInt((Integer) o);
-      } else if (t == DataTypes.LongType) {
+      } else if (pdt instanceof PhysicalIntegerType) {
+        if (o instanceof Date) {
+          dst.appendInt(DateTimeUtils.fromJavaDate((Date) o));
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java:141]: This is a bit weird. Why the caller side passes `Date` for date type but `Long` for timestamp type?

Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java:144]: https://github.com/apache/spark/blob/5d90f9825bc5aad1cf24b773586cba5309ee6caa/sql/catalyst/src/main/scala/org/apache/spark/sql/Row.scala#L170-L174

Except for `java.sql.Timestamp ', I think we should also handle `o` is `java.time.LocalDate` and `java.time.Instant` ref the document.?

There should be no tests to cover these branches. I will add some test later.

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java [hunk 3]
```diff
@@ -125,32 +125,45 @@ public static Map<Integer, Integer> toJavaIntMap(ColumnarMap map) {
   }
 
   private static void appendValue(WritableColumnVector dst, DataType t, Object o) {
```
Comment (cloud-fan) [inline sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java:127]: If this function is to handle external types (Row), I think matching logical data type makes more sense, as external types are bound to logical types.

Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java:127]: OK, I will revert the change of this method, but there maybe 2 problem:

- TimestampType is not handled, will throw `UnsupportedOperationException` directly
- appendValue method not handle MapType yet

seems that another pr is needed to solve these?

Comment (cloud-fan) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java]: Yea, this is an existing bug and should be fixed separately.

**Review by cloud-fan — APPROVED** [review]

**Comment by LuciferYang** [comment]
GA passed

**Comment by cloud-fan** [comment]
thanks, merging to master!

File: sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java (hunk 3 above)
Comment (LuciferYang) [reply sql/core/src/main/java/org/apache/spark/sql/execution/vectorized/ColumnVectorUtils.java]: > Yea, this is an existing bug and should be fixed separately.

thanks @cloud-fan,  I will follow up on the remaining bug fix later

**Pull request closed without merging**